### ISSUE TITLE
kubectl port-forward broken pipe

### ISSUE THREAD

**Issue by dwbrown2** [kubernetes/kubernetes#74551]
**What happened**:

We have a pod running in our k8s cluster that I connect to via `kubectl port-forward`. I am able to connect to this pod (using the following command) but then start getting broken pipe error messages after maintaining a connection for what is typically 30-60s. 

`kubectl port-forward --namespace monitoring deployment/cost-monitor 9090`

There seems to be a correlation between error rate and the amount of data being transferred. I see the following error message initially:

`E0225 15:20:06.212139   26392 portforward.go:363] error copying from remote stream to local connection: readfrom tcp6 [::1]:9090->[::1]:57794: write tcp6 [::1]:9090->[::1]:57794: write: broken pipe`

These errors are oftentimes followed by timeout messages but necessarily immediately:

`E0225 15:22:30.454203   26392 portforward.go:353] error creating forwarding stream for port 9090 -> 9090: Timeout occured`

**What you expected to happen**:
No error messages or major degradation in transfer rate.

**How to reproduce it (as minimally and precisely as possible)**:
Connect via port-forward and transfer ~5 mb over several minutes. 

**What else?**:
We have multiple HTTP requests being made at any given time on this port-forward connection. 

**Environment**:
Experiencing on both AWS kops and GKE 

// kops
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-04T04:48:03Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.11", GitCommit:"637c7e288581ee40ab4ca210618a89a555b6e7e9", GitTreeState:"clean", BuildDate:"2018-11-26T14:25:46Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"linux/amd64"}

// GKE
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-04T04:48:03Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"10+", GitVersion:"v1.10.11-gke.1", GitCommit:"5c4fddf874319c9825581cc9ab1d0f0cf51e1dc9", GitTreeState:"clean", BuildDate:"2018-11-30T16:18:58Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}

**Comment by AjayTripathy** [comment]
Hi, I’m hitting this on my cluster too.

Here's what we're currently running:
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.3", GitCommit:"721bfa751924da8d1680787490c54b9179b1fed0", GitTreeState:"clean", BuildDate:"2019-02-04T04:48:03Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.6", GitCommit:"b1d75deca493a24a2f87eb1efde1a569e52fc8d9", GitTreeState:"clean", BuildDate:"2018-12-16T04:30:10Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}

Any additional repro info that I can provide?

**Comment by ysimonson** [comment]
I'm seeing this issue as well. I think kubelet is not properly closing the error channel associated with port forwarded connections, because I'm seeing the port forwarder get stuck [here](https://github.com/kubernetes/kubernetes/blob/1cb550295a9314ace7eb2df1f0d1cf7578fc11ba/staging/src/k8s.io/client-go/tools/portforward/portforward.go#L398).

You don't even need to send a lot of data, just break the reading end of the connection from the client-side while the port forwarder is trying to write. Assuming you're port forwarding an HTTP server available on port 30600 locally, this will reliably reproduce the issue on k8s for docker for mac (server version v1.10.11):

```
#!/usr/bin/env python3

import datetime
import http.client

for _ in range(10):
    conn = http.client.HTTPConnection("localhost:30600")
    conn.request("GET", "/")
    conn.getresponse()
    conn.close()
    print(datetime.datetime.now())

```

I can reliably reproduce the issue on a newer version of k8s + ubuntu + minikube as well, but it is more resilient, and the above script won't trigger the problem. I'm not sure yet how to make a minimally reproducible test for that target yet.

**Comment by sandeepsharip** [comment]
What's the fix for this

**Comment by kargakis** [comment]
I have created a simple client+server reproducer at https://github.com/kargakis/k8s-74551-reproducer

This issue is unfortunate, `kubectl port-forward` makes it very easy for our less kube-experienced developers to leverage our production deployment for development.

**Comment by montanaflynn** [comment]
I also get this, especially when using kubectl port-forward for redis or postgres or haproxy.

```
E0807 23:29:49.561205   87544 portforward.go:303] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:5432->127.0.0.1:52434: write tcp4 127.0.0.1:5432->127.0.0.1:52434: write: broken pipe
```

**Comment by rihardsk** [comment]
I get the same issue when trying to access an in-cluster docker registry. When i try to push an image right after the forwarding is started, everything's fine for about 3 seconds - a couple of layers manage to get pushed but then the upload quickly grinds to a halt and never completes.

Here's output from the `kubectl port-forward`
```sh
$ kubectl port-forward --namespace kube-system deployment/kube-registry-v0 5000:5000
Forwarding from 127.0.0.1:5000 -> 5000
Forwarding from [::1]:5000 -> 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
Handling connection for 5000
E0816 15:40:24.189095   21402 portforward.go:400] an error occurred forwarding 5000 -> 5000: error forwarding port 5000 to pod a732546ab28635579a54b6e4ab964ea8e5f1f2da8d71f3ee3aaa37ee06b4cb6f, uid : exit status 1: 2019/08/16 12:40:24 socat[8459] E write(6, 0x22fbd70, 8192): Broken pipe
Handling connection for 5000
Handling connection for 5000
E0816 15:40:38.593757   21402 portforward.go:362] error creating forwarding stream for port 5000 -> 5000: Timeout occured
Handling connection for 5000
```
And here's how the `docker push` output looks when it has halted
```sh
$ docker push localhost:5000/taskservice:latest
The push refers to repository [localhost:5000/taskservice]
430da0e622b8: Layer already exists 
cc133a8f03f4: Layer already exists 
fed3b16d3013: Layer already exists 
a49039162690: Layer already exists 
0c3a922281a0: Layer already exists 
54577ffe58e1: Layer already exists 
dd6d39e085f7: Layer already exists 
a0ec25fc3a30: Retrying in 1 second 
dfb8f512e5cc: Pushed 
3ca254258b97: Pushing [==================================================>]  34.24MB
abba6879d6f2: Retrying in 1 second 
d9c78f8e26b3: Retrying in 1 second 
a82245a38cb3: Waiting 
085245055bb7: Waiting 
24f93a4970a3: Waiting 
e79142719515: Waiting 
aeda103e78c9: Waiting 
2558e637fbff: Waiting 
f749b9b0fb21: Waiting 
```

If stop the port forwarding process (which, btw, doesn't respond to Ctrl-C, I have to do `kill -9`) and retry the whole process the same thing happens again (and i manage to upload another few layers).

**Comment by tallskog** [comment]
Also got the same kind of problem as @rihardsk. When looking at the kube-registry logs from the container, at least the following error logs were seen

`time="2019-08-18T14:28:00Z" level=error msg="response completed with error" err.code="blob unknown" err.detail=sha256:bbab4ec87ac4f89eaabdf68dddbd1dd930e3ad43bded38d761b89abf9389a893 err.message="blob unknown to registry" go.version=go1.6.3 http.request.host="localhost:5000" http.request.id=969f3ee4-aa19-4821-9d74-0e5a7639deba http.request.method=HEAD http.request.remoteaddr="127.0.0.1:51570" http.request.uri="/v2/sdltestapp/blobs/sha256:bbab4ec87ac4f89eaabdf68dddbd1dd930e3ad43bded38d761b89abf9389a893" http.request.useragent="docker/19.03.1 go/go1.12.5 git-commit/74b1e89 kernel/3.10.0-957.27.2.el7.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.1 \\(linux\\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=98.755941ms http.response.status=404 http.response.written=157 instance.id=be7e4ed5-99dd-49b8-93b6-e5dbbd0c2aac vars.digest="sha256:bbab4ec87ac4f89eaabdf68dddbd1dd930e3ad43bded38d761b89abf9389a893" vars.name=sdltestapp version=v2.5.1 
127.0.0.1 - - [18/Aug/2019:14:28:00 +0000] "HEAD /v2/sdltestapp/blobs/sha256:bbab4ec87ac4f89eaabdf68dddbd1dd930e3ad43bded38d761b89abf9389a893 HTTP/1.1" 404 157 "" "docker/19.03.1 go/go1.12.5 git-commit/74b1e89 kernel/3.10.0-957.27.2.el7.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.1 \\(linux\\))"

time="2019-08-18T14:28:01Z" level=error msg="unknown error reading request payload: write /var/lib/registry/docker/registry/v2/repositories/sdltestapp/_uploads/840c6e4f-4c28-4b24-ad52-be8753612e39/data: cannot allocate memory" go.version=go1.6.3 http.request.host="localhost:5000" http.request.id=ecf2db39-cbfa-4b52-ad0b-af6b839d931e http.request.method=PATCH http.request.remoteaddr="127.0.0.1:51458" http.request.uri="/v2/sdltestapp/blobs/uploads/840c6e4f-4c28-4b24-ad52-be8753612e39?_state=ebA5Em9IUSedYo_D2I8p31EGKwkIdRzlXfnwPR5U0CB7Ik5hbWUiOiJzZGx0ZXN0YXBwIiwiVVVJRCI6Ijg0MGM2ZTRmLTRjMjgtNGIyNC1hZDUyLWJlODc1MzYxMmUzOSIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxOS0wOC0xOFQxNDoyNzo0NS4wODA2NDk3MDlaIn0%3D" http.request.useragent="docker/19.03.1 go/go1.12.5 git-commit/74b1e89 kernel/3.10.0-957.27.2.el7.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.1 \\(linux\\))" instance.id=be7e4ed5-99dd-49b8-93b6-e5dbbd0c2aac vars.name=sdltestapp vars.uuid=840c6e4f-4c28-4b24-ad52-be8753612e39 version=v2.5.1 

time="2019-08-18T14:28:03Z" level=error msg="response completed with error" err.code=unknown err.detail="write /var/lib/registry/docker/registry/v2/repositories/sdltestapp/_uploads/840c6e4f-4c28-4b24-ad52-be8753612e39/data: cannot allocate memory" err.message="unknown error" go.version=go1.6.3 http.request.host="localhost:5000" http.request.id=ecf2db39-cbfa-4b52-ad0b-af6b839d931e http.request.method=PATCH http.request.remoteaddr="127.0.0.1:51458" http.request.uri="/v2/sdltestapp/blobs/uploads/840c6e4f-4c28-4b24-ad52-be8753612e39?_state=ebA5Em9IUSedYo_D2I8p31EGKwkIdRzlXfnwPR5U0CB7Ik5hbWUiOiJzZGx0ZXN0YXBwIiwiVVVJRCI6Ijg0MGM2ZTRmLTRjMjgtNGIyNC1hZDUyLWJlODc1MzYxMmUzOSIsIk9mZnNldCI6MCwiU3RhcnRlZEF0IjoiMjAxOS0wOC0xOFQxNDoyNzo0NS4wODA2NDk3MDlaIn0%3D" http.request.useragent="docker/19.03.1 go/go1.12.5 git-commit/74b1e89 kernel/3.10.0-957.27.2.el7.x86_64 os/linux arch/amd64 UpstreamClient(Docker-Client/19.03.1 \\(linux\\))" http.response.contenttype="application/json; charset=utf-8" http.response.duration=17.792958322s http.response.status=500 http.response.written=212 instance.id=be7e4ed5-99dd-49b8-93b6-e5dbbd0c2aac vars.name=sdltestapp vars.uuid=840c6e4f-4c28-4b24-ad52-be8753612e39 version=v2.5.1 
`

**Comment by kargakis** [comment]
@kubernetes/sig-api-machinery-bugs

**Comment by rihardsk** [comment]
Increasing the memory limit for the registry container to 200Mi seems to have helped me as well. Thanks @tallskog!

**Comment by kwlzn** [comment]
we're also seeing similar issues at Twitter in our usages of `kubectl port-forward`.

**Comment by dims** [comment]
/cc

long-term-issue (note to self)

**Comment by lavalamp** [comment]
I think this evaded our triage process because the api machinery label was put on quite a while after it was created.

**Comment by dwbrown2** [comment]
@lavalamp is there additional information that would be useful for getting this triaged?

**Comment by Multiply** [comment]
I can also reliably reproduce it, but trying to fetch a 6MB file over a simple http server.

**Comment by cforce** [comment]
Sadly but true, get rid of tiller a root for this issue by using helm3

**Comment by lavalamp** [comment]
I tagged this with help-wanted; I think the next step is for someone to try @kargakis's [reproducer](https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-490988621) and figure out if the problem is kubectl doing something wrong or apiserver doing something wrong.

**Comment by mnaser** [comment]
Can anyone share the architecture behind how they are running the Kubernetes API server?  I suspect we're dealing with this issue because we're running it behind `haproxy` and there is some sort of idle timeout at play.

**Comment by mnaser** [comment]
Upon adding the following to our `haproxy` front-end, this issue has pretty much been eliminated.

```
timeout client 4h
```

and the following to the backend:

```
timeout server 4h
```

Upon making that change, absolutely zero timeouts for an idle `port-forward` for quite sometime.  There must be similar options if you have a different load balancer in front of your Kubernetes API.  Perhaps this is something that people who are affected by it can try that.

**Comment by kargakis** [comment]
I doubt this is an external LB issue, I was able to reproduce in a local cluster.

**Comment by mnaser** [comment]
@kargakis Fair enough, I actually would never have this issue with local clusters (but would consistently run against remote ones), so this bug might still be a thing but maybe some of us affected actually are running into a external load balancer issue and thought I'd put my notes there anyways.

**Comment by timkessler** [comment]
I am running into the timeout issue consistently after port-forward hands off 27 connections to start streaming to the same port (psql connections to postgres). I get intermittent failures on connections  after the first 27.

**Comment by ndarilek** [comment]
I've got a slightly different error, and am wondering if it is this issue or another:

```
E0122 10:09:20.251054  560781 portforward.go:385] error copying from local connection to remote stream: read tcp6 [::1]:8080->[::1]:36774: read: connection reset by peer
E0122 10:09:20.252475  560781 portforward.go:372] error copying from remote stream to local connection: readfrom tcp6 [::1]:8080->[::1]:36776: write tcp6 [::1]:8080->[::1]:36776: write: connection reset by peer
E0122 10:09:50.321506  560781 portforward.go:340] error creating error stream for port 8080 -> 8080: Timeout occured
```

I.e. connection reset, not broken pipe. The pods in question continue to run just fine. Happens when I try serving a ~70 MB file from a local Minikube cluster. Accessing the ingress directly via `minikube ip` works perfectly.

If this is another issue, please point me to it so I can a) see if it is fixed and b) update our local development tooling accordingly. Thanks.

**Comment by drx777** [comment]
FWIW, I ran into 

    E0130 12:37:06.971972   32446 portforward.go:400] an error occurred forwarding 10001 -> 9090: error forwarding port 9090 to pod <someid>, uid : exit status 1: 2020/01/30 11:37:06 socat[23094] E write(5, 0x560b3b7332f0, 7): Broken pipe

And this was resolved by using plaintext (`usePlaintext()`) on the channel.

**Comment by lavalamp** [comment]
>  `usePlaintext()`

Can you say more about that? That's suspicious, if the problem is binary data, maybe there's some specific byte sequence that is tickling a bug in the proxy.

**Comment by drx777** [comment]
> Can you say more about that? That's suspicious, if the problem is binary data, maybe there's some specific byte sequence that is tickling a bug in the proxy.

This is grpc communication. The channel creation code goes like this:

     val channel = ManagedChannelBuilder
            .forAddress(grpcListenAddress, grpcPort)
            // .usePlaintext()
            .directExecutor().build()

And the error I get is

    io.grpc.StatusRuntimeException: UNAVAILABLE: io exception
    Channel Pipeline: [SslHandler#0, ProtocolNegotiators$ClientTlsHandler#0, WriteBufferingAndExceptionHandler#0, DefaultChannelPipeline$TailContext#0]

When I enable the `.usePlaintext()` line, there's no problem with the tunneling itself anymore.

**Comment by lovexiongmao** [comment]
@dwbrown2 have you solve the problem? i get the same issue,now. and It affects my production environment

**Comment by dwbrown2** [comment]
No great solutions here... we are just using port forwarding a lot less for the time being :(

**Comment by Chaya56** [comment]
Got the same, it's working 10 sec then it fail:
`Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
E0217 11:52:07.206492   18833 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:39112: write tcp4 127.0.0.1:8080->127.0.0.1:39112: write: broken pipe
Handling connection for 8080
Handling connection for 8080
E0217 11:52:07.242230   18833 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:39114: write tcp4 127.0.0.1:8080->127.0.0.1:39114: write: broken pipe
E0217 11:52:07.246288   18833 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:39118: write tcp4 127.0.0.1:8080->127.0.0.1:39118: write: broken pipe
Handling connection for 8080
E0217 11:52:07.446730   18833 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:39130: write tcp4 127.0.0.1:8080->127.0.0.1:39130: write: broken pipe
Handling connection for 8080
E0217 11:52:37.180892   18833 portforward.go:340] error creating error stream for port 8080 -> 8080: Timeout occured
E0217 11:52:37.209572   18833 portforward.go:340] error creating error stream for port 8080 -> 8080: Timeout occured
E0217 11:52:37.271464   18833 portforward.go:340] error creating error stream for port 8080 -> 8080: Timeout occured
E0217 11:52:37.283550   18833 portforward.go:362] error creating forwarding stream for port 8080 -> 8080: Timeout occured
E0217 11:52:37.447476   18833 portforward.go:340] error creating error stream for port 8080 -> 8080: Timeout occured`

kubectl & cluster version:
`Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-18T23:30:10Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"16", GitVersion:"v1.16.4", GitCommit:"224be7bdce5a9dd0c2fd0d46b83865648e2fe0ba", GitTreeState:"clean", BuildDate:"2019-12-11T12:37:43Z", GoVersion:"go1.12.12", Compiler:"gc", Platform:"linux/amd64"}
`

**Comment by Chaya56** [comment]
Upgrading my cluster to 1.17 seems to fix my issue. Now kubectl and server are in the same version

**Comment by LighthouseJ** [comment]
I setup a redis cluster w/ helm chart from here: https://github.com/helm/charts/tree/master/stable/redis  I'm using the "production" values and leaving password enabled.  I'm using NiFi outside of kubernetes to send it data, and using port-forward (for now) to connect.

One difference is that when I use clients like redis-cli and Redis Desktop Manager, then they work fine, but when I use NiFi, then it fails consistently with auth failures despite repeatedly ensuring the right redis password is used.

This NiFi flow has worked in the past with a standalone Redis instance not in kubernetes too.  No password was used in this first case BTW.

```
E0318 12:41:19.922991   26692 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:6379->127.0.0.1:48768: read: connection reset by peer
Handling connection for 6379
E0318 12:41:20.927939   26692 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:6379->127.0.0.1:48770: read: connection reset by peer
Handling connection for 6379
E0318 12:41:21.933259   26692 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:6379->127.0.0.1:48772: read: connection reset by peer
Handling connection for 6379
```

```
2020-03-18 12:46:21,591 WARN [Timer-Driven Process Thread-5] o.a.n.controller.tasks.ConnectableTask Administratively Yielding PutDistributedMapCache[id=d4415025-0170-1000-02ba-32d1b5cc21b9] due to uncaught Exception: org.springframework.data.redis.RedisConnectionFailureException: Cannot get Jedis connection; nested exception is redis.clients.jedis.exceptions.JedisConnectionException: Could not get a resource from the pool
org.springframework.data.redis.RedisConnectionFailureException: Cannot get Jedis connection; nested exception is redis.clients.jedis.exceptions.JedisConnectionException: Could not get a resource from the pool
	at org.springframework.data.redis.connection.jedis.JedisConnectionFactory.fetchJedisConnector(JedisConnectionFactory.java:281)
...
Caused by: redis.clients.jedis.exceptions.JedisDataException: ERR invalid password
	at redis.clients.jedis.Protocol.processError(Protocol.java:127)
	at redis.clients.jedis.Protocol.process(Protocol.java:161)
	at redis.clients.jedis.Protocol.read(Protocol.java:215)
	at redis.clients.jedis.Connection.readProtocolWithCheckingBroken(Connection.java:340)
```
I'm using redis 5.0.7, kubectl, etc... v 1.17.3, w/ minikube 1.8.2 on a Virtualbox VM that is also running kubernetes v 1.17.3 too, so they should be in complete agreement.

**Comment by lavalamp** [comment]
OK. I strongly suspect a bug that is triggered by some particular binary sequence.

It might be here: https://github.com/kubernetes/kubernetes/blob/v1.19.0-alpha.0/staging/src/k8s.io/apimachinery/pkg/util/httpstream/spdy/connection.go

but more likely is in the github.com/docker/spdystream dependency.

**Comment by lavalamp** [comment]
I don't see a random data transmission test there (???!??). https://github.com/docker/spdystream/blob/master/spdy_bench_test.go#L68 might be easy to adapt if someone has time to investigate.

**Comment by anks333** [comment]
I am facing the same issue, 
error logs
```
Handling connection for 80
E0429 23:06:59.964905   90120 portforward.go:400] an error occurred forwarding 80 -> 88: error forwarding port 88 to pod d4110d7f601d61db997a6f16c670f283f66ec544aba15b87c010c1b68bbe555f, uid : exit status 1: 2020/04/29 17:36:59 socat[8709] E connect(5, AF=2 127.0.0.1:88, 16): Connection refused
```

**Comment by aojea** [comment]
I was going through the issue and I don't have clear that all the errors commented here are SPDY problems, some of the errors mentioned in that issue seems network related issues, some extract from different comments:

* `socat[8709] E connect(5, AF=2 127.0.0.1:88, 16): Connection refused`
* `read: connection reset by peer`
* `write: broken pipe Handling connection for 8080`
* `socat[23094] E write(5, 0x560b3b7332f0, 7): Broken pipe`

> but then start getting broken pipe error messages after maintaining a connection for what is typically 30-60s

This may be a known problem with TCP connections that become idle and traverse different devices that keep status, but is weird since you say that it happens while transmitting traffic.

Is there a load balancer in front of the apiserver?
Can you verify if the connections are not closed if they are idle by any element in between?

**Comment by montanaflynn** [comment]
@aojea the original issue and myself both have the error:

`error copying from remote stream to local connection ... write: broken pipe`

Full errors:

```
// https://github.com/kubernetes/kubernetes/issues/74551
E0225 15:20:06.212139 26392 portforward.go:363] error copying from remote stream to local connection: readfrom tcp6 [::1]:9090->[::1]:57794: write tcp6 [::1]:9090->[::1]:57794: write: broken pipe

// https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-519361205
E0807 23:29:49.561205  87544 portforward.go:303] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:5432->127.0.0.1:52434: write tcp4 127.0.0.1:5432->127.0.0.1:52434: write: broken pipe
```

I think the other commented errors should be addressed in a separate issue so we could nail down why this specific error happens and fix it. 

I get the error when I `port-forward` directly to haproxy, redis or postgres pod and try to read a lot of data through it. There is no load balancer in front of the apiserver. We're using AWS EKS with version `v1.13.12-eks-eb1860` and client version `v1.16.3`

**Comment by aojea** [comment]
I've tried to reproduce it locally without success, but I don't know if I'm missing something, I've used iperf instead of the other reproducer https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-490988621 to avoid possibles issues with golang or the custom implenentation, iperf is very mature in this area ;).

These are the steps I followed:

1. Create a multinode cluster with kind (kubernetes 1.18.2)
2. Create a deployment with iperf server running on it
```
kubectl run iperf --command --image=us.gcr.io/k8s-artifacts-prod/e2e-test-images/agnhost:2.13 --port=8001 -- /bin/sh -c "/usr/local/bin/iperf -s -p 8001"
```
3. Forward a port locally to the pod
```
kubectl port-forward deployment/iperf 8888:8001
Forwarding from 127.0.0.1:8888 -> 8001
Forwarding from [::1]:8888 -> 8001

Handling connection for 8888
Handling connection for 8888
```

4. Run an iperf client using the forwarded port during 600s

```
iperf -c localhost -p 8888 -i 1 -t 600
------------------------------------------------------------
Client connecting to localhost, TCP port 8888
TCP window size: 2.50 MByte (default)
------------------------------------------------------------
[  3] local 127.0.0.1 port 33982 connected with 127.0.0.1 port 8888
[ ID] Interval       Transfer     Bandwidth
[  3]  0.0- 1.0 sec   433 MBytes  3.63 Gbits/sec
...
[  3] 598.0-599.0 sec   475 MBytes  3.99 Gbits/sec
[  3] 599.0-600.0 sec   458 MBytes  3.85 Gbits/sec
[  3]  0.0-600.0 sec   273 GBytes  3.91 Gbits/sec
```


There are no broken pipe, it will be great if I can reproduce it locally.
@montanaflynn can you try the iperf test in your environment?

**Comment by montanaflynn** [comment]
@aojea so I actually just tried to reproduce the issue with haproxy, postgres and redis but was unable to get the same error as before. 

We did upgrade our kubernetes from 1.12.x to 1.13.x since I originally commented, not sure if that fixed it as another user mentioned upgrading fixed the problem for them as well.

Didn't have any errors with iperf either, if possible I would suggest trying it on kubernetes 1.12.x.

**Comment by ysimonson** [comment]
I haven't seen this issue in a while either, whereas it came up on a daily basis about a year ago. I think it was fixed somewhere between v1.10.11 and v1.15.5.

**Comment by aojea** [comment]
I'm afraid that the 1.12.x is not going to receive any backport, the last supported version is 1.16

https://kubernetes.io/docs/setup/release/version-skew-policy/

If this fixed in 1.16 and later versions I think that this issue should be closed, but as I said before, I'm happy to investigate if somebody can point me to a way to reproduce it in a supported version

**Comment by evgenmedia** [comment]
Hello everybody! Got the same error when run helm in tiller-deploy pod :(

E0501 19:13:40.230413      14 portforward.go:400] an error occurred forwarding 42309 -> 44134: error forwarding port 44134 to pod 03a12c5d4dd5160ad83ed45813ec706afd9309c905884aaeccca818c03cbdcf6, uid : exit status 1: 2020/05/01 22:13:40 socat[32022] E write(5, 0x5628c782eea0, 24): Broken pipe

root@master-1:~# kubectl version 
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:56:40Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.2", GitCommit:"52c56ce7a8272c798dbc29846288d7cd9fbae032", GitTreeState:"clean", BuildDate:"2020-04-16T11:48:36Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}


It's fresh installed K8S cluster on bare metal with 1 master & 2 workers nodes.

**Comment by morhook** [comment]
With [astronomer helm chart](https://github.com/astronomer/airflow-chart), `ubuntu` + `minikube` this error happens a lot!

```
kubectl port-forward svc/airflow-webserver 8080:8080 -n airflow
```

**Comment by reinhaun** [comment]
I also get this. Running Airflow on EKS
`Client Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:26:26Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"15+", GitVersion:"v1.15.11-eks-af3caf", GitCommit:"af3caf6136cd355f467083651cc1010a499f59b1", GitTreeState:"clean", BuildDate:"2020-03-27T21:51:36Z", GoVersion:"go1.12.17", Compiler:"gc", Platform:"linux/amd64"}`

**Comment by morhook** [comment]
For avoiding this problem I’m creating a NodePort service and then using the command `minikube service service-name -n airflow`. Of course this can be done as my minikube is my private cluster.

**Comment by 3nueves** [comment]
Hi everyone!
I have bare-metal kubernetes cluster (v1.18.6 ) with kubeadm and I have the same problem:

````
$kubectl port-forward svc/argocd-server -n argocd 8080:443
Forwarding from 127.0.0.1:8080 -> 8080
Forwarding from [::1]:8080 -> 8080
Handling connection for 8080
Handling connection for 8080
Handling connection for 8080
E0730   8542 portforward.go:372] error copying from remote stream to local connection: readfrom tcp6 [::1]:8080->[::1]:36698: write tcp6 [::1]:8080->[::1]:36698: write: broken pipe
Handling connection for 8080
E0730    8542 portforward.go:233] lost connection to pod
````
The errors I get from the kubernetes nodes are:
````
sudo journalctl -xe -f
.
.
kubelet[1643]: E0730   1643 httpstream.go:251] error forwarding port 8080 to pod 1e8e65b01cedddfd259e6036210cd401a8679d3313eb80179060fee35cc96bd5, uid : EOF:
```` 

Does anyone know anything about this topic?

**Comment by bayudwiyansatria** [comment]
Hello got same issues running with `stacked etcd cluster` behind external Load Balancer. 
Using `nginx` or `haproxy` got same.

kubectl version 

```
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:58:53Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.6", GitCommit:"dff82dc0de47299ab66c83c626e08b245ab19037", GitTreeState:"clean", BuildDate:"2020-07-15T16:51:04Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by seperman** [comment]
Minikube + minio runs into this issue all the time when streaming data.

Kubernetes version:

```
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-20T12:43:34Z", GoVersion:"go1.13.9", Compiler:"gc", Platform:"linux/amd64"
```

**Comment by moonek** [comment]
Similar errors often occur on nodes where tiller is deployed using helm v2.
Are there any correct causes and actions?

```
Sep 07 10:49:04 worker-02 kubelet[12563]: E0907 10:49:04.314685   12563 httpstream.go:251] error forwarding port 44134 to pod 213bee5d1092c9f93aa7f5298646e3a77b782abb7005a6dcdb8a3a5f45f167a6, uid : EOF:
Sep 07 17:52:38 worker-02 kubelet[12563]: E0907 17:52:38.449474   12563 httpstream.go:251] error forwarding port 44134 to pod 213bee5d1092c9f93aa7f5298646e3a77b782abb7005a6dcdb8a3a5f45f167a6, uid : write tcp 127.0.0.1:46101->127.0.0.1:55378: write: broken pipe:
```

**Comment by lavalamp** [comment]
This issue is waiting for a contributor to dig in and diagnose. My current guess is recorded above [in the comment here](https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-600750593).

**Comment by nfisher** [comment]
wonder if there is an easy way to get pcap running at various points in the connection. kubectl to api should be ok if the key exchange can be dumped. I guess the other point would be api to kube-proxy? Would that be direct or routed through the api servers kube-proxy and then to the pods node kube-proxy? I assume pod to kube-proxy would be ok or pod networking in k8s would be seeing similar connectivity issues.

**Comment by hdiass** [comment]
I'm having same issue on AKS kubernetes v1.16.13. I use port-forwarding to connect to mongodb and i face this issue when doing backup from a local mongodump connecting to a remote mongodb via k8s cluster.

**Comment by igor9silva** [comment]
I'm facing it constantly with a MongoDB pod. Seems to happen more often with concurrent connections (I'm doing a `mongodump`).

```
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.8", GitCommit:"9f2892aab98fe339f3bd70e3c470144299398ace", GitTreeState:"clean", BuildDate:"2020-08-14T00:06:38Z", GoVersion:"go1.13.15", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by archonic** [comment]
This is happening with the `linkerd dashboard` command as well. The localhost connection to the dashboard fails and the terminal connection to the linkerd dashboard severs, so it just puts random errors into your terminal and you can't reconnect to the dashboard.

**Comment by VGerris** [comment]
hi,

I have something that might be related:

E1027 15:47:19.719563   26356 portforward.go:400] an error occurred forwarding 8082 -> 8080: error forwarding port 8080 to pod , uid : Cannot connect to the Docker daemon at unix:///var/run/docker.sock. Is the docker daemon running?

This happens when we stress test over a tunnel towards a service deployed in GKE.
From 2000 request sent by a node.js script, only 1494 came through and 506 of this message were logged by the port forward console that was open (on a Windows 10 client machine).

Would be nice if we can send a bit more over the wire.
Is this a bug in the go script, docker, or is it some OS related setting that causes this?

**Comment by cypherfox** [comment]
> This is happening with the `linkerd dashboard` command as well. The localhost connection to the dashboard fails and the terminal connection to the linkerd dashboard severs, so it just puts random errors into your terminal and you can't reconnect to the dashboard.

I think I have a similar problem with the linkerd cli. Could you please provide the vesions of:
* linkerd cli
* linkerd control plane
* kubectl 
* kubernetes cluster

**Comment by velmafia** [comment]
I have similar problem, this occurs in azure k8s cluster(client: v1.19.3, server: v1.17.3) and in local `kind` cluster (client: v1.19.3, server: v1.19.1).
It's looks very similar like @ysimonson suggestion [here](https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-476418180). And (I guess) its related with http streams.

Problem occurs when you stop downloading webpage/file during the downloading is in progress. Then `broken pipe` logs appears, and after a while `Timeout occured`. After that, everything stops, and connection is unusable.

```bash
Forwarding from 127.0.0.1:8080 -> 80
Handling connection for 8080
Handling connection for 8080
E1105 16:49:04.050790   90091 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:8080->127.0.0.1:56050: read: connection reset by peer
E1105 16:49:04.051655   90091 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:56050: write tcp4 127.0.0.1:8080->127.0.0.1:56050: write: broken pipe
Handling connection for 8080
E1105 16:49:04.351923   90091 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:8080->127.0.0.1:56052: read: connection reset by peer
Handling connection for 8080
E1105 16:49:04.354772   90091 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:56052: write tcp4 127.0.0.1:8080->127.0.0.1:56052: write: broken pipe
E1105 16:49:04.712995   90091 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:8080->127.0.0.1:56058: read: connection reset by peer
Handling connection for 8080
E1105 16:49:04.715697   90091 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:8080->127.0.0.1:56058: write tcp4 127.0.0.1:8080->127.0.0.1:56058: write: broken pipe
E1105 16:49:34.715510   90091 portforward.go:340] error creating error stream for port 8080 -> 80: Timeout occured
```

I don't exactly know how k8s/kublet works in details, but this logs from `containerd` always appear when connection is broken.
```
Nov 05 18:57:04 kind-control-plane containerd[48]: E1105 18:57:04.975425      48 httpstream.go:143] (conn=&{0xc0004f7e40 [0xc000b70b40 0xc000b70be0 0xc0005903c0 0xc000b70c80 0xc000b70d20] {0 0} 0x55fe64dbd9d0}, request=2) timed out waiting for streams
```
When problem occurs in "production" this logs have a hundreds of connections inside `[]`.

But, long story short:
I prepared [reproduction of this bug](https://github.com/velmafia/k8s_issue-74551) with github actions, and problem occurs in workflow: https://github.com/velmafia/k8s_issue-74551/actions/runs/346450615 
`kind` logs are stored in artifacts, feel free to download/fork and retest also with your version on k8s.

**Comment by tuskiomi** [comment]
Can confirm that issue exists in latest build.

**Comment by aojea** [comment]
are you able to reproduce using iperf as explained here?
https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-622412024

**Comment by tuskiomi** [comment]
> are you able to reproduce using iperf as explained here?
> [#74551 (comment)](https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-622412024)

no, I'm able to re-create it with a remote cluster, however. I am able to keep a normal ssh stream up with the cluster, but not a port-forward.

**Comment by aojea** [comment]
one of the reasons, there can be more, is that the connection can be idle and some intermediate devices with low timers close the TCP session, so it eventually times out in one/both the endpoints of the connection.
SSH can configure periodic keepalives to keep the traffic flowing and renewing the timeouts, or maybe you are using the session ...
This is why is important to know if with a continuous TCP stream you are hitting the issue,

**Comment by anthcor** [comment]
Just noticed this myself with a fresh single node installation.

```sh
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:20:00Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

Port forwarding didn't work at first

```sh
$ kubectl -n argo port-forward deployment/argo-server 2746:2746
Forwarding from 127.0.0.1:2746 -> 2746
Forwarding from [::1]:2746 -> 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
E0128 10:49:21.896317   13138 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:2746->127.0.0.1:65062: write tcp4 127.0.0.1:2746->127.0.0.1:65062: write: broken pipe
Handling connection for 2746
E0128 10:49:51.866614   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
E0128 10:49:52.849451   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
Handling connection for 2746
E0128 10:50:22.857444   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
Handling connection for 2746
E0128 10:50:52.863231   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
Handling connection for 2746
Handling connection for 2746
E0128 10:51:22.904116   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
E0128 10:51:22.904268   13138 portforward.go:340] error creating error stream for port 2746 -> 2746: Timeout occured
Handling connection for 2746
Handling connection for 2746
^C%
```

When I let kubectl decide the host port, port-forwarding worked!

```sh
$ kubectl -n argo port-forward deployment/argo-server :2746
Forwarding from 127.0.0.1:65199 -> 2746
Forwarding from [::1]:65199 -> 2746
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
Handling connection for 65199
^CHandling connection for 65199
```

Then forwarding to the same port on the host worked!

```sh
$ kubectl -n argo port-forward deployment/argo-server 2746:2746
Forwarding from 127.0.0.1:2746 -> 2746
Forwarding from [::1]:2746 -> 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
Handling connection for 2746
```

**Comment by xkortex** [comment]
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"20", GitVersion:"v1.20.2", GitCommit:"faecb196815e248d3ecfb03c680a4507229c2a56", GitTreeState:"clean", BuildDate:"2021-01-13T13:28:09Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.4", GitCommit:"d360454c9bcd1634cf4cc52d1867af5491dc9c5f", GitTreeState:"clean", BuildDate:"2020-11-11T13:09:17Z", GoVersion:"go1.15.2", Compiler:"gc", Platform:"linux/amd64"}
```

I am also trying to spin up Argo on a fresh minikube instance. It was working fine yesterday, had to tear everything down due to other issues, when I follow the [Argo Quick-start](https://github.com/argoproj/argo-workflows/blob/master/docs/quick-start.md) followed by [Access token creation](https://argoproj.github.io/argo-workflows/access-token/). Started with trying to `curl https://localhost:2746/api/v1/workflows/argo -H "Authorization: $ARGO_TOKEN"`, it didn't like my cert (`curl: (60) SSL certificate problem: self signed certificate`), so I tried with `--insecure`

```
{"code":14,"message":"connection error: desc = \"transport: authentication handshake failed: x509: certificate relies on legacy Common Name field, use SANs or temporarily enable Common Name matching with GODEBUG=x509ignoreCN=0\""}%
```

```
$kubectl -n argo port-forward deployment/argo-server 2746:2746

Forwarding from 127.0.0.1:2746 -> 2746
Forwarding from [::1]:2746 -> 2746
Handling connection for 2746
E0303 15:38:45.260055 3581829 portforward.go:385] error copying from local connection to remote stream: read tcp4 127.0.0.1:2746->127.0.0.1:48246: read: connection reset by peer
Handling connection for 2746
...
```

So then I tried athcor's trick: 

```shell
$ kubectl -n argo port-forward deployment/argo-server :2746

Forwarding from 127.0.0.1:46691 -> 2746
Forwarding from [::1]:46691 -> 2746
```

I'm pretty sure I'm able to curl `http` (no s) `localhost:2746` and get the Argo gui index, as well as the `/api/v1/version` endpoint, or even just `/api/v1` and get a json "no method" response. 
```
$ curl http://localhost:2746
curl: (7) Failed to connect to localhost port 2746: Connection refused
```

I switched back to `kubectl -n argo port-forward deployment/argo-server 2746:2746` and now it seems to be `curl: (52) Empty reply from server`. 

I destroyed the namespace, started over, and now it seems to be behaving, albeit I'm still getting permissions from the API, at least I'm getting valid JSON responses back.

**Comment by aidinhass** [comment]
I am not sure how much this is relevant to this thread but try https://localhost:2746 instead of 127.0.0.1:2746.

**Comment by ReillyBrogan** [comment]
@aidinhass It's not relevant at all as `port-forward` has to do with TCP proxying and has nothing to do with HTTPS. 

If you are interested in reading more about this you should read up about the [OSI model](https://en.wikipedia.org/wiki/OSI_model) as this issue is a [layer 4](https://en.wikipedia.org/wiki/Transport_layer) issue. HTTPS is a [layer 7](https://en.wikipedia.org/wiki/Application_layer) protocol.

**Comment by Apokalypser** [comment]
I have the same problem pushing to a port-forwarded registry on a bare-metal Kubernetes cluster. Letting kubectl choose the local port seems to work around it.
@xkortex You need to execute curl http://localhost:46691 after using that workaround ;-)

**Comment by jaroli** [comment]
Same problem for Grafana on AWS EKS.
```
~> kubectl run graf --image=grafana/grafana
~> kubectl port-forward graf 3000:3000
Forwarding from 127.0.0.1:3000 -> 3000
Forwarding from [::1]:3000 -> 3000
Handling connection for 3000
Handling connection for 3000
E0426 22:52:17.142466    1736 portforward.go:372] error copying from remote stream to local connection: readfrom tcp4 127.0.0.1:3000->127.0.0.1:63479: write tcp4 127.0.0.1:3000->127.0.0.1:63479: write: broken pipe
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
Handling connection for 3000
```
```
~> kubectl version
Client Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.0", GitCommit:"e19964183377d0ec2052d1f1fa930c4d7575bd50", GitTreeState:"clean", BuildDate:"2020-08-26T14:30:33Z", GoVersion:"go1.15", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"19+", GitVersion:"v1.19.8-eks-96780e", GitCommit:"96780e1b30acbf0a52c38b6030d7853e575bcdf3", GitTreeState:"clean", BuildDate:"2021-03-10T21:32:29Z", GoVersion:"go1.15.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by mcallaghan-geotab** [comment]
The "trick" above did not help.

Attempts to send `5MB` payloads consistently dies after 3x attempts.
```
$ kubectl port-forward deployment/blah-some-deployment-in-gke :5000
Forwarding from 127.0.0.1:44731 -> 5000
Forwarding from [::1]:44731 -> 5000
Handling connection for 44731
E0526 15:42:20.300993   28563 portforward.go:400] an error occurred forwarding 44731 -> 5000: error forwarding port 5000 to pod 95446003a7fe376f92b8c37225615041e36688f340097fede14c7a93a0e60d66, uid : exit status 1: 2021/05/26 19:42:20 socat[689383] E write(5, 0x558a42e31e20, 8192): Broken pipe
Handling connection for 44731
E0526 15:42:22.297885   28563 portforward.go:400] an error occurred forwarding 44731 -> 5000: error forwarding port 5000 to pod 95446003a7fe376f92b8c37225615041e36688f340097fede14c7a93a0e60d66, uid : exit status 1: 2021/05/26 19:42:22 socat[689395] E write(5, 0x5652032b9e20, 8192): Broken pipe
Handling connection for 44731
E0526 15:42:23.152925   28563 portforward.go:400] an error occurred forwarding 44731 -> 5000: error forwarding port 5000 to pod 95446003a7fe376f92b8c37225615041e36688f340097fede14c7a93a0e60d66, uid : exit status 1: 2021/05/26 19:42:23 socat[689400] E write(5, 0x55c860f5ce20, 8192): Broken pipe
Handling connection for 44731
E0526 15:42:53.622741   28563 portforward.go:340] error creating error stream for port 44731 -> 5000: Timeout occurred
```

**Comment by eladm26** [comment]
happens to me when port-forwarding to minio and trying to download large files, getting:
`E0706 18:38:06.048525       1 portforward.go:340] error creating error stream for port 9000 -> 9000: Timeout occured`

**Comment by waeljammal** [comment]
Happens to me as well with services like redis. Running K3D on a ubuntu server vm.

**Comment by mnp** [comment]
I had success with @anthcor 's solution: let him decide the local port. 

```shell
kubectl  port-forward svc/foo :8080
Forwarding from 127.0.0.1:57708 -> 8080
Forwarding from [::1]:57708 -> 8080
Handling connection for 57708
Handling connection for 57708 ... etc```

I ALSO had success by specifying local address as 127.0.0.1 (I don't need ipv6).   The WEIRD thing is after doing this I can go back to the original form,  `kubectl  port-forward svc/foo 1234:8080` and it works again. This smells like a socket reuse issue.

This is docker-desktop on a mac.
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.2", GitCommit:"092fbfbf53427de67cac1e9fa54aaa09a28371d7", GitTreeState:"clean", BuildDate:"2021-06-16T12:59:11Z", GoVersion:"go1.16.5", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"19", GitVersion:"v1.19.7", GitCommit:"1dd5338295409edcfff11505e7bb246f0d325d15", GitTreeState:"clean", BuildDate:"2021-01-13T13:15:20Z", GoVersion:"go1.15.5", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by leontastic** [comment]
I had a similar problem using `kubectl port-forward` and I resolved it with `ulimit -n 65536` on Mac OS.

**Comment by pratikpatil024** [comment]
Hi @leontastic , can you please explain what you did?

Let's say I am running the following command to forward the port:
```
kubectl port-forward pod/ca-c4b68bbbc-6l7wh 7054:7054
```

How can I use `ulimit` here to achieve the same goal?

Thanks!

**Comment by leontastic** [comment]
I ran `ulimit -n 65536` as-is. You might need `sudo` on your system. This increases the file descriptor limit of the local shell where you're running `kubectl port-forward`.

My hypothesis is that `kubectl port-forward` doesn't clean up its sockets properly so the local shell runs into the file descriptor limit after some time under high load (or maybe after terminating a few times). This seemed to stop port-forward from breaking all the time when I was running automated tests against a K8s service I was trying to debug.

**Comment by chymy** [comment]
I have a similar problem in kubelet logs:
```
Sep 14 10:55:03 os4-control-plane-zlt92.novalocal kubelet[7038]: E0914 10:55:03.380811    7038 httpstream.go:251] error forwarding port 2379 to pod 4c33311464fed819cad31e0f9ef84e0b29a8ff09ebe2b6034fb003780f65ad12, uid : exit status 1: 2021/09/14 10:55:03 socat[9371] E write(1, 0x1411420, 24): Broken pipe
Sep 14 10:55:03 os4-control-plane-zlt92.novalocal kubelet[7038]: E0914 10:55:03.521429    7038 httpstream.go:251] error forwarding port 2379 to pod 4c33311464fed819cad31e0f9ef84e0b29a8ff09ebe2b6034fb003780f65ad12, uid : EOF:
Sep 14 10:55:05 os4-control-plane-zlt92.novalocal kubelet[7038]: E0914 10:55:04.999702    7038 httpstream.go:251] error forwarding port 2379 to pod 4c33311464fed819cad31e0f9ef84e0b29a8ff09ebe2b6034fb003780f65ad12, uid : EOF:
```

```
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.806253   23954 httpstream.go:45] Upgrading port forward response
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.806378   23954 httpstream.go:53] (conn=0xc000373050) setting port forwarding streaming connection idle timeout to 4h0m0s
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.806397   23954 httpstream.go:209] (conn=0xc000373050) waiting for port forward streams
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.811080   23954 httpstream.go:245] (conn=0xc000373050, request=0) invoking forwarder.PortForward for port 2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.811928   23954 docker_streaming_others.go:55] executing port forwarding command: /usr/bin/nsenter -t 32445 -n /usr/bin/socat - TCP4:localhost:2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.858772   23954 httpstream.go:45] Upgrading port forward response
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.858902   23954 httpstream.go:53] (conn=0xc0008382a0) setting port forwarding streaming connection idle timeout to 4h0m0s
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.858922   23954 httpstream.go:209] (conn=0xc0008382a0) waiting for port forward streams
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.863238   23954 httpstream.go:245] (conn=0xc0008382a0, request=0) invoking forwarder.PortForward for port 2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.864134   23954 docker_streaming_others.go:55] executing port forwarding command: /usr/bin/nsenter -t 32445 -n /usr/bin/socat - TCP4:localhost:2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.882546   23954 httpstream.go:247] (conn=0xc0008382a0, request=0) done invoking forwarder.PortForward for port 2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: E0914 11:13:08.882595   23954 httpstream.go:251] error forwarding port 2379 to pod 4c33311464fed819cad31e0f9ef84e0b29a8ff09ebe2b6034fb003780f65ad12, uid : EOF:
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: I0914 11:13:08.887124   23954 httpstream.go:247] (conn=0xc000373050, request=0) done invoking forwarder.PortForward for port 2379
Sep 14 11:13:08 os4-control-plane-zlt92.novalocal kubelet[23954]: E0914 11:13:08.887187   23954 httpstream.go:251] error forwarding port 2379 to pod 4c33311464fed819cad31e0f9ef84e0b29a8ff09ebe2b6034fb003780f65ad12, uid : EOF:
```

Is there any solution?

**Comment by RaananHadar** [comment]
I did not have success with @anthcor 's workaround. Hope this gets fixed.

**Comment by bhechinger** [comment]
@anthcor's workaround won't work for me sadly so I'd really like to see this addressed. The weird thing is I have less trouble when I use `kubectl port-forward` than when I use `k8s.io/client-go/tools/portforward` directly.

**Comment by BnJam** [comment]
@anthonycorletti 's fix works for me using K8s v1.21 
Letting the host choose the port alone has less error messages, but still enough to be problematic. But then setting it back to a fixed port after cancelling the previous step has resukting very good performance and connection. 

I did use CleanMyMacX inbetween those steps which cleared a lot of caches and the DNS cache - probably unrelated..

**Comment by anthonycorletti** [comment]
https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-910520361 is the best short-term fix imo. have run into this many times testing many services port forwarding to my local network

**Comment by horihel** [comment]
that short-term fix doesn't work on windows though - or does it?

**Comment by atinsinghal97** [comment]
I use port forward to interact with applications like ArgoCD (web ui) and have been using this as a workaround in case folks want to try: https://github.com/atinsinghal97/kubectl-port-forward-heaven

**Comment by victor-fialkin-deltatre** [comment]
In my case mapping to a different local port worked (9000:9021)

**Comment by hspens** [comment]
Changing the **mtu** of the network interface to match that of the docker network solved the problem for me. 

```
> ifconfig | grep mtu
# tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1300
``` 

```
> sudo ifconfig tun0 mtu 1500 up
> ifconfig | grep mtu
# tun0: flags=4305<UP,POINTOPOINT,RUNNING,NOARP,MULTICAST>  mtu 1500
```

**Comment by varunbpatil** [comment]
I'm having this issue with applications that expose a web-UI like temporal, pgadmin, etc. The only thing that works for me is to have a completely unrelated upload running in the background continuously. I know, doesn't make sense to me either, but leaving something like speedtest-cli running in a loop in a terminal makes kubectl port-forward work without any issues and much more snappier than before.

**Comment by msa21** [comment]
Check your pod (endpoints) linked with the service you are doing port -forward for . and then check whether the pod is listening on 80 and 443 port number both if yes then http and https both request will work and if pod is only listening on port 80 then please do not use https request.

**Comment by maxmarcon** [comment]
I'm facing similar issues to the ones described here.

K8S versions:
```
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.11", GitCommit:"5824e3251d294d324320db85bf63a53eb0767af2", GitTreeState:"clean", BuildDate:"2022-06-16T05:33:01Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"21+", GitVersion:"v1.21.12-eks-a64ea69", GitCommit:"d4336843ba36120e9ed1491fddff5f2fec33eb77", GitTreeState:"clean", BuildDate:"2022-05-12T18:29:27Z", GoVersion:"go1.16.15", Compiler:"gc", Platform:"linux/amd64"}
```
OS: MacOS 12.4

I'm connecting to a EKS K8S cluster. If I run `kubectl port-forward` with a high verbosity flag, I see that before starting listening for connections the protocol is upgraded to SPDY:

```
Protocols in 376 milliseconds
I0622 18:55:03.749811    3295 round_trippers.go:460] Response Headers:
I0622 18:55:03.749816    3295 round_trippers.go:463]     Date: Wed, 22 Jun 2022 16:55:03 GMT
I0622 18:55:03.749820    3295 round_trippers.go:463]     Connection: Upgrade
I0622 18:55:03.749823    3295 round_trippers.go:463]     Upgrade: SPDY/3.1
I0622 18:55:03.749826    3295 round_trippers.go:463]     X-Stream-Protocol-Version: portforward.k8s.io
```
 
So port-forward uses SPDY under the hood? I wonder if this could be related to the problem.

Other than that, I tried different things but I'm at a loss. Any hint on how to solve the issue would be really appreciated.

**Comment by deepaksinghvi** [comment]
Destroying the namespace and recreated it helped in this case. 
It was ok to do it on local but not practical in other environments.

**Comment by freebased** [comment]
same here with kube-state-metrics:
```
Handling connection for 8080
E1018 13:40:03.813851  132449 portforward.go:391] error copying from local connection to remote stream: read tcp4 127.0.0.1:8080->127.0.0.1:45398: read: connection reset by peer
```

**Comment by trym** [comment]
FWIW: Able to reproduce using https://github.com/0xmichalis/k8s-74551-reproducer on v1.25.0. If I replace the kubectl port-forward with a port forward using rust/kube-rs, the forwarder reports broken pipe errors, but the the requests are successfully served. This seems to indicate that kubectl does not deal with the broken pipe errors properly? Disclaimer: only spent a few minutes looking into this.

**Comment by sxllwx** [comment]
I'm trying to locate and fix this issue 【WIP】

---
environment:

```
kubectl, kube-apiserver, kubelet, all from master, commit 1f0e718585af775ba847f6fc36fa287e3f2ecd19
containerd:  v1.6.19 1e1ea6e986c6c86565bc33d52e34b81b3e2bc71f
```
---

thx for   https://github.com/kargakis/k8s-74551-reproducer
---

【Please correct me if I'm wrong】

Here are some screenshots for troubleshooting:

--- 

kubectl port-forward general process:

<img width="745" alt="image" src="https://user-images.githubusercontent.com/11829907/233098634-60cf7ea3-3d5b-44d7-b322-36fc28c3aa40.png">

---

pcap in the target container namespace: 

<img width="1493" alt="image" src="https://user-images.githubusercontent.com/11829907/233098743-391bb182-37c1-41d2-91d1-d8f5ecbdc009.png">

--- 

reproducible results：

<img width="1476" alt="image" src="https://user-images.githubusercontent.com/11829907/233099027-afa0797a-0d6d-48fe-b68a-1aa6e743e501.png">


--- 

The reason identified so far is: 

- In the screenshot, it is obvious that containerd receives RST
- containerd writes the problems encountered in the process of forwarding traffic into the stream, and finally feeds back to our client, `that is kubectl.`


---
So I made some fixes to containerd, https://github.com/sxllwx/containerd/commit/28755fff7d67b64576054e4fbc4845e116d92b63ji

I'm running the failed example fine with containerd using this patch.

**Comment by dims** [comment]
Looks like it's the last commit in this branch: https://github.com/sxllwx/containerd/commits/fix/k8s-issue-74551

**Comment by lavalamp** [comment]
Wow! I wasn't expecting that culprit! great find!

**Comment by morhook** [comment]
Thanks for finding this problem guys and gals! Hope to be able to do port-forward on my clusters when containerd is deployed and upgraded!

**Comment by sxllwx** [comment]
Sorry, disappeared for a week [I had a very busy week], update the latest progress:

> 💡 I figured out the magic of https://github.com/0xmichalis/k8s-74551-reproducer

code segment:

https://github.com/0xmichalis/k8s-74551-reproducer/blob/master/server. go#L11

The http-server just prints the Method and ContentLength of http, and returns it directly **without** reading the request.Body. This will cause the data in the kernel buffer to be unable to be read to the user state [causing subsequent WindowFull, ZeroWindow, and subsequent RST]. When the Body of the Request is very large, the buffer will be quickly filled. [The client simulates an HTTP request with a very large request body].

> Why do we reset the Stream when EPIPE occurs instead of the previous Close logic?

- Stream Reset, reset will close the connection, and then trigger the release of Select here
- Stream Close will simply remove the httpstream and will not affect the connection, causing the long-term blocking of the Select here to cause the client to hang.
      
https://github.com/containerd/containerd/blob/f84b2b47b0faca2 23837fbdb489961d15b023119/pkg/ cri/streaming/portforward/httpstream.go#L219

**Comment by sxllwx** [comment]
> Maybe you have the same question as me: We see in pcap that the process sends RST to containerd, why we don’t get `read: connection reset by peer` error in the Read direction, but always receive `write : broken pipe`Error:

The answer is here: Carefully, we can see that the pcap indicates that the process sent a FIN before sending the RST. Therefore, io.Copy can end normally in the code segment.
https://github.com/containerd/containerd/blob/290a800e83d5460207cf198516359bfd2b5038d6/pkg/cri/server/sandbox_portforward_linux.go#L84-L89

**But when the RST arrives, a `write: broken pipe` error occurs in the write direction.**

----
So far, I think the background of the whole problem is very clear. If you have any questions, you can communicate here at any time

**Comment by thesuperzapper** [comment]
@sxllwx thanks for your work! Do you think that this issue https://github.com/kubernetes/kubectl/issues/1368, is also caused by the same problem you identified?

**Comment by sxllwx** [comment]
> @sxllwx thanks for your work! Do you think that this issue [kubernetes/kubectl#1368](https://github.com/kubernetes/kubectl/issues/1368), is also caused by the same problem you identified?

Looks like the same problem. You can try to use the containerd branch of https://github.com/sxllwx/containerd/tree/fix/k8s-issue-74551, I believe this issue can be resolved.

**Comment by giuliohome** [comment]
I see this error message while getting a redis .net library exception: is the issue here or there?

https://github.com/StackExchange/StackExchange.Redis/issues/2470

**Comment by vishwaspai** [comment]
I've had some success in getting better reliability by reducing MTU of the network interface.

**Comment by hadifarnoud** [comment]
is there any fix without upgrading containerd? I do not have access

**Comment by kong62** [comment]
minio error:
```
# kubectl -n openobserve port-forward --address 0.0.0.0 svc/openobserve-minio-console 9001:9001
Forwarding from 0.0.0.0:9001 -> 9001
Handling connection for 9001
E1201 17:46:43.211933  584240 portforward.go:381] error copying from remote stream to local connection: readfrom tcp4 10.64.57.72:9001->172.16.49.86:4023: write tcp4 10.64.57.72:9001->172.16.49.86:4023: write: broken pipe
Handling connection for 9001
E1201 17:47:13.219367  584240 portforward.go:347] error creating error stream for port 9001 -> 9001: Timeout occurred
Handling connection for 9001
```

**Comment by npuichigo** [comment]
any update here?

**Comment by sxllwx** [comment]
> any update here?

xref: https://github.com/kubernetes/kubernetes/issues/107203

**Comment by sxllwx** [comment]
> it looks like you might be working on a fix (based on your commits), is that correct?

Yes, this problem is caused by the streaming api of the container's runtime (for example: containerd, cri-o). The implementation of the runtime API relies on (in the form of fork or vendor) https://github.com/kubernetes/kubernetes/blob/dc3f5ec6ccb9855dfa99f4c1078625df5fdfab6a/staging/src/k8s.io/kubelet/pkg/cri/streaming/portforward/ httpstream.go

So we fixed it and added relevant e2e tests . https://github.com/kubernetes/kubernetes/pull/117493

According to [Mr. aojea's comments](https://github.com/kubernetes/kubernetes/pull/117493#issuecomment-1972017769), this problem is expected to be fixed in the future. Thanks to @aojea and @brianpursley for their work. 😄

**Comment by calmacx** [comment]
So this is still a problem on the latest release? 
I'm getting it:
```
error copying from remote stream to local connection: readfrom tcp6 [::1]:3000->[::1]:61249: write tcp6 [::1]:3000->[::1]:61249: write: broken pipe
```
Im on version `1.31.4`

**Comment by montanaflynn** [comment]
Nearly 6 years after [my original comment in this issue](https://github.com/kubernetes/kubernetes/issues/74551#issuecomment-519361205) I still run into this a lot.

I think I've found an okay solution, at least for my needs. Forwarding the port to the pod directly instead of going through the service has been the solution to my connection issues

```
# lots of broken pipes and connection issues going through the service
kubectl port-forward svc/redis 6379:6379 

# connecting to the pod directly, no issues
kubectl port-forward redis-0 6379:6379 
```

**Comment by rafagsiqueira** [comment]
I am guessing the PR was never merged?
